packages/blockchain/examples/4444.ts

```diff
@@ -0,0 +1,85 @@
+import { createBlock } from '@ethereumjs/block'
+import { createBlockchain } from '@ethereumjs/blockchain'
+import { Common, Hardfork, Mainnet } from '@ethereumjs/common'
+import { bytesToHex, hexToBytes } from '@ethereumjs/util'
+
+const main = async () => {
+  const common = new Common({ chain: Mainnet, hardfork: Hardfork.Shanghai })
+  // Use the safe static constructor which awaits the init method
+  const blockchain = await createBlockchain({
+    common,
+  })
+
+  // We are using this to create minimal post merge blocks between shanghai and cancun in line with the
+  // block hardfork configuration of mainnet
+  const chainTTD = BigInt('58750000000000000000000')
+  const shanghaiTimestamp = 1681338455
+
+  // We use minimal data to provide a sequence of blocks (increasing number, difficulty, and then setting parent hash to previous block)
+  const block1 = createBlock(
+    {
+      header: {
+        // 15537393n is terminal block in mainnet config
+        number: 15537393n + 500n,
+        // Could be any parenthash other than 0x00..00 as we will set this block as a TRUSTED 4444 anchor
+        // instead of genesis to build blockchain on top of. One could use any criteria to set a block
+        // as trusted 4444 anchor
+        parentHash: hexToBytes(`0x${'20'.repeat(32)}`),
+        timestamp: shanghaiTimestamp + 12 * 500,
+      },
+    },
+    { common, setHardfork: true },
+  )
+  const block2 = createBlock(
+    {
+      header: {
+        number: block1.header.number + 1n,
+        parentHash: block1.header.hash(),
+        timestamp: shanghaiTimestamp + 12 * 501,
+      },
+    },
+    { common, setHardfork: true },
+  )
+  const block3 = createBlock(
+    {
+      header: {
+        number: block2.header.number + 1n,
+        parentHash: block2.header.hash(),
+        timestamp: shanghaiTimestamp + 12 * 502,
+      },
+    },
+    { common, setHardfork: true },
+  )
+
+  let headBlock, blockByHash, blockByNumber
+  headBlock = await blockchain.getCanonicalHeadBlock()
+  console.log(
+    `Blockchain ${blockchain.consensus.algorithm} Head: ${headBlock.header.number} ${bytesToHex(headBlock.hash())}`,
+  )
+  // Blockchain casper Head: 0 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3
+
+  // 1. We can put any post merge block as 4444 anchor by using TTD as parentTD
+  // 2. For pre-merge blocks its prudent to supply correct parentTD so as to respect the
+  //    hardfork configuration as well as to determine the canonicality of the chain on future putBlocks
+  await blockchain.putBlock(block1, { parentTd: chainTTD })
```

Comment:
I still do not fully understand this anchor concept. What is anchored and how?

Reply:
block1 is the 4444 anchor i.e. the parent chain to block1 doesn't need to be added to the blockchain and blockchain can be build forward from here

it acts as a re-genesis anchor, one specifies an anchor by providing parentTD, for a PoS block parentTD can just be chainTTD, for a PoW block an actual parentTD should be added so as to eventually and correctly matchup with the terminal block TTD when forward chain blocks will be added on top of it

Reply:
These are one of the things which are too implicit, at least on the API, maybe the DB level. I am still not fully grasping if the `parentTd` is also needed (for calculating/validating something) or is only set in some way to indicate the anchor status.

No matter how it is: this is unnecessarily implicit and we should think about the "anchor" (or call it checkpoint?) status and give this its own API and database representation.

I think "Checkpoint" might actually be the more common and established terminolgy.

So I think we definitely want to have some external relation in the DB where we store all the checkpoints (and not only have this in some very implicit form), and can e.g. extend our API with things like:

```ts
blockchain.getCheckpoints()
blockchain.isCheckpoint(5n)
```

(or first there would be the question if there can be only one of these, but guess there can be several respectively would make sense?)

(wonder if we can even lign in the genesis block in this concept, guess this is exactly the same as a checkpoint?)

And then for setting these checkpoints, this should be very explicitly named. So if `parentTd` is still necessary, even if we store a `Checkpoints -> [ Block Numbers ]` relation, then this should minimally be provided with naming the option `setCheckpoint: 560934n` (and so provide the parentTD still, but set the very emphasis on this checkpoint setting, so one can read from the API call what is happening here).